code/game/machinery/dance_machine.dm

```diff
@@ -453,30 +483,46 @@
 	if(world.time < stop && active)
 		var/sound/song_played = sound(selection.song_path)
 
-		for(var/mob/M in range(10,src))
-			if(!M.client || !(M.client.prefs.read_preference(/datum/preference/toggle/sound_jukebox)))
+		// Goes through existing mobs in rangers to determine if they should not be played to
+		for(var/datum/weakref/weak_to_hide_from as anything in rangers)
+			var/mob/to_hide_from = weak_to_hide_from?.resolve()
+			if(QDELETED(to_hide_from) \
+				|| get_dist(src, get_turf(to_hide_from)) > 10 \
+				|| !to_hide_from.client \
+				|| !(to_hide_from.client.prefs.read_preference(/datum/preference/toggle/sound_jukebox)) \
+			)
+				rangers -= weak_to_hide_from
+				to_hide_from?.stop_sound_channel(CHANNEL_JUKEBOX)
+
+		// Collect mobs to play the song to, stores weakrefs of them in rangers
+		for(var/mob/to_play_to in range(world.view, src))
+			if(!to_play_to.client || !(to_play_to.client.prefs.read_preference(/datum/preference/toggle/sound_jukebox)))
```

Comment:
```suggestion
			if(isnull(to_play_to.client) || !(to_play_to.client.prefs.read_preference(/datum/preference/toggle/sound_jukebox)))
```

might need to be multlined? maybe we could just have a local macro that's something like so we don't have to keep having these long ass `if()` chains. small potential example that may or may not work

```dm
#define HAS_JUKEBOX_PREF(mob) (!QDELETED(mob) && !isnull(mob.client) && mob.client.prefs.read_preference(/datum/preference/toggle/sound_jukebox)
```

up to you though, just mentioning it since i noticed this pattern repeating a lot through this PR